### PR TITLE
gh-117180: Complete call sequence when trace stack overflow

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
* Issue: gh-117180

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
What was happening is that the trace stack overflowed, so it ended up deopting in the middle of a call sequence, which has inconsistent stack effect with its tier 1 equivalent, causing the interpreter to segfault. The optimizer just caught that earlier.

File: Python/optimizer.c [hunk 1]
```diff
@@ -475,8 +475,11 @@ BRANCH_TO_GUARD[4][2] = {
 #define TRACE_STACK_PUSH() \
     if (trace_stack_depth >= TRACE_STACK_SIZE) { \
         DPRINTF(2, "Trace stack overflow\n"); \
-        OPT_STAT_INC(trace_stack_overflow); \
-        ADD_TO_TRACE(_EXIT_TRACE, 0, 0, 0); \
+        OPT_STAT_INC(trace_stack_overflow);      \
+        /* Remove the call sequence */           \
+        for (int x = 0; x < i; x++) {            \
```
Comment (gvanrossum) [inline Python/optimizer.c]: I'm uncomfortable with this referencing `i` (it's easy to overlook). Can you make it a parameter of the macro?

File: Python/optimizer.c [hunk 2]
```diff
@@ -475,8 +475,11 @@ BRANCH_TO_GUARD[4][2] = {
...
+           trace[trace_length-x].opcode = _NOP;  \
+        } \
```
Comment (gvanrossum) [inline Python/optimizer.c]: OTOH a correct fix that keeps the `_PUSH_FRAME` call we've been working up to would be this:
```suggestion
        ADD_TO_TRACE(uop, oparg, operand, target); \
        ADD_TO_TRACE(_EXIT_TRACE, 0, 0, 0); \
```
i.e. just emit the `_PUSH_FRAME` uop -- it has the correct target. (I would restore the original spacing of the `OPT_STAT_INC()` macro too.)

**Review by gvanrossum — CHANGES_REQUESTED** [review]
I now realize I broke this in gh-117028 when I moved the `ADD_TO_TRACE()` calls around to get prettier debug output (in particular, with the correct `operand` printed); I tried to be careful to make sure there was always exactly one `ADD_TO_TRACE()` in each path through the code, but I messed up -- I missed the `goto done` hidden in the  `TRACE_STACK_PUSH()` macro.

We should probably have tests for all these paths through the code. Thanks again to @lazorchakp for finding this!

File: Python/optimizer.c [hunk 3]
```diff
@@ -475,7 +475,8 @@ BRANCH_TO_GUARD[4][2] = {
 #define TRACE_STACK_PUSH() \
     if (trace_stack_depth >= TRACE_STACK_SIZE) { \
         DPRINTF(2, "Trace stack overflow\n"); \
-        OPT_STAT_INC(trace_stack_overflow); \
+        OPT_STAT_INC(trace_stack_overflow);      \
```
Comment (gvanrossum) [inline Python/optimizer.c]:
```suggestion
        OPT_STAT_INC(trace_stack_overflow); \
```

**Review by gvanrossum — APPROVED** [review]
Thanks!

**Pull request merged**